cmd/esc/cli/cli_test.go

```diff
@@ -635,6 +639,122 @@ func (c *testPulumiClient) ListEnvironmentRevisions(
 	return resp, nil
 }
 
+// CreateEnvironmentRevisionTag creates a new revision tag with the given name.
+func (c *testPulumiClient) CreateEnvironmentRevisionTag(
+	ctx context.Context,
+	orgName string,
+	envName string,
+	tagName string,
+	revision *int,
+) error {
+	env, _, err := c.getEnvironment(orgName, envName, "")
+	if err != nil {
+		return err
+	}
+
+	rev := len(env.revisions)
+	if revision != nil {
+		rev = *revision
+	}
+	if rev < 1 || rev > len(env.revisions) {
+		return errors.New("not found")
+	}
+
+	if _, ok := env.tags[tagName]; ok {
+		return errors.New("already exists")
+	}
+
+	env.tags[tagName] = rev
+	return nil
+}
+
+// GetEnvironmentRevisionTag returns a description of the given revision tag.
+func (c *testPulumiClient) GetEnvironmentRevisionTag(
+	ctx context.Context,
+	orgName string,
+	envName string,
+	tagName string,
+) (*client.EnvironmentRevisionTag, error) {
+	env, _, err := c.getEnvironment(orgName, envName, "")
+	if err != nil {
+		return nil, err
+	}
+
+	rev, ok := env.tags[tagName]
+	if !ok {
+		return nil, &apitype.ErrorResponse{Code: http.StatusNotFound}
+	}
+	return &client.EnvironmentRevisionTag{Name: tagName, Revision: rev}, nil
+}
+
+// UpdateEnvironmentRevisionTag updates the revision tag with the given name.
+func (c *testPulumiClient) UpdateEnvironmentRevisionTag(
+	ctx context.Context,
+	orgName string,
+	envName string,
+	tagName string,
+	revision *int,
+) error {
+	env, _, err := c.getEnvironment(orgName, envName, "")
+	if err != nil {
+		return err
+	}
+
+	rev := len(env.revisions)
+	if revision != nil {
+		rev = *revision
+	}
+	if rev < 1 || rev > len(env.revisions) {
+		return errors.New("not found")
+	}
+
+	if _, ok := env.tags[tagName]; !ok {
```

Comment:
Should this just create it in this case?

Reply:
No--that would differ from the behavior of the ESC API.

Reply:
(this code is the test-only implementation of the REST client, which fakes up the ESC API)

Reply:
Should we consider changing the ESC API to handle an upsert case?

Reply:
Worth considering, yeah. It'd save us an API call and improve the clarity a bit.

Reply:
(I'm going to leave it out of scope for this PR, though)